source/server/listener_manager_impl.cc

```diff
@@ -393,6 +393,17 @@ bool ListenerManagerImpl::addOrUpdateListenerInternal(
   auto existing_active_listener = getListenerByName(active_listeners_, name);
   auto existing_warming_listener = getListenerByName(warming_listeners_, name);
 
+  // The listener should be updated back to its original state and the warming listener should be
+  // removed.
+  if (existing_warming_listener != warming_listeners_.end() &&
+      existing_active_listener != active_listeners_.end() &&
+      (*existing_active_listener)->blockUpdate(hash)) {
```

Comment:
I have to look into blockUpdate before I asked how to verify the active listener and the latest config is equivalent. Sigh...

Reply:
`blockUpdate()` just checks that the hash of the two listeners is the same
https://github.com/envoyproxy/envoy/blob/51551ae944c642e6fc61563cbea8653087e70f1f/source/server/listener_impl.h#L256

https://github.com/envoyproxy/envoy/blob/51551ae944c642e6fc61563cbea8653087e70f1f/source/common/protobuf/utility.cc#L264-L277

It's used just below this snippet to ignore duplicate listeners: https://github.com/envoyproxy/envoy/blob/25fcc8b40523d74c28495208cae2cc4cf701d98b/source/server/listener_manager_impl.cc#L409-L415